### PR TITLE
Hierarchical metadata@main

### PR DESCRIPTION
**Pull request by BFalquet**
Add functions to combine rules or list of rules.

thank you for the review

### CODE REVIEW

File: R/rules.R [hunk 1]
```diff
@@ -120,3 +120,96 @@ as.list.rule <- function(x, ...) {
...
+#'   ),
+#'   r3 = rule(
+#'     SOMETHING = "something"
+#'   )
+#' )
+#'
+#' combineListRules(l1, l2)
+combineListRules <- function(x, val, ...) {
+  checkmate::assert_list(x, types = "rule", null.ok = FALSE, names = "named")
+  checkmate::assert_list(val, types = "rule", null.ok = FALSE, names = "named")
```
Comment (clarkliming) [inline R/rules.R:205]: names = "unique" to ensure there are no duplicates

File: NAMESPACE [hunk 2]
```diff
@@ -21,6 +21,8 @@ export(assert_valid_list_format)
 export(attr_label)
 export(attr_label_df)
 export(co_relevels)
+export(combineListRules)
```
Comment (clarkliming) [inline NAMESPACE:24]: let's keep consistent naming conventions, we always use underscore instead of CamelCase

File: NEWS.md [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 * Rules specified under the `all_datasets` keyword in a format list will apply to every data set of the reformatted object unless specified otherwise.
 * New `verbose` argument in the `reformat` method. When applied to `list` the value of this augment can be controlled with the `dunlin.reformat.verbose` option or the `R_DUNLIN_REFORMAT_VERBOSE` environment variable.
+* New `combine_rules` and `combineListRules.Rd` functions to combine rules or list of rules into a single rule or a single list of rules.
```
Comment (clarkliming) [inline NEWS.md:5]: `.Rd` suffix can be unnecessary

File: R/rules.R [hunk 4]
```diff
@@ -120,3 +120,96 @@ as.list.rule <- function(x, ...) {
 
   r_list
 }
+
+#' Combine Two Rules
+#'
+#' @param x (`rule`) to modify.
+#' @param y (`rule`) rule whose mapping will take precedence over the ones described in `x`.
```
Comment (clarkliming) [inline R/rules.R:127]: the later one takes priority. seems this is quite normal and we also should use it in citril right

File: tests/testthat/test-rules.R [hunk 5]
```diff
@@ -116,3 +116,115 @@ test_that("as.list and rule are reversible when .to_NA is NULL", {
...
+
+# combine_rules ----
+
+test_that("combine_rules works as expected", {
+  r1 <- rule(a = "1", b = "2", .to_NA = "x", .drop = TRUE, .na_last = FALSE)
+  r2 <- rule(a = "3", c = "4", .to_NA = "y", .drop = FALSE)
+
+  res <- combine_rules(r1, r2)
+  expect_s3_class(res, "rule")
+  expect_identical(res, rule(a = "3", c = "4", b = "2", .to_NA = "y", .drop = FALSE, .na_last = TRUE))
```
Comment (clarkliming) [inline tests/testthat/test-rules.R:128]: why the order is `a`, `c`, and `b`?

Comment (clarkliming) [reply tests/testthat/test-rules.R:128]: should we, natually have `a` and `b` first (from rule 1), and then other elements in rule 2?

Comment (BFalquet) [reply tests/testthat/test-rules.R:128]: I decided to place first the mapping from the rule that has the priority (here rule2). We could have an option to apply an alphabetic sorting.

File: R/rules.R (hunk 4 above)
Comment (BFalquet) [reply R/rules.R:127]: Yes, here it is more natural since we are used to `modifyList`.

File: tests/testthat/test-rules.R (hunk 5 above)
Comment (BFalquet) [reply tests/testthat/test-rules.R:128]: Do you think we should go the other way around @clarkliming ?

File: R/rules.R [hunk 6]
```diff
@@ -120,3 +120,96 @@ as.list.rule <- function(x, ...) {
...
+#'   .drop = TRUE
+#' )
+#' r2 <- rule(
+#'   "first" = c("F", "f"),
+#'   "second" = c("S", "s"),
+#'   "third" = c("T", "t"),
+#'   .to_NA = "something"
+#' )
+#' combine_rules(r1, r2)
+combine_rules <- function(x, y, safe = TRUE, ...) {
```
Comment (clarkliming) [inline R/rules.R:150]: is there any situation that `safe` is necessary?

File: R/rules.R [hunk 7]
```diff
@@ -125,3 +125,96 @@ as.list.rule <- function(x, ...) {
...
+
+  # If one of the rules is NULL, return the other (via empty list).
+  x <- as.list(x)
+  y <- as.list(y)
+  names_y <- names(y)
+  names_x <- setdiff(names(x), names(y))
+
+  x <- x[names_x]
+  r <- c(y, x)
+  r <- do.call(rule, r)
```
Comment (clarkliming) [inline R/rules.R:172]: shall we keep the order of `x` instead of using the order of `y`

Comment (BFalquet) [reply R/rules.R:172]: ok, we can do that and append the levels of `y` at the end.

File: R/rules.R (hunk 6 above)
Comment (BFalquet) [reply R/rules.R]: actually not necessary at the moment, I removed it and left `...` to future-proof the function.

File: R/rules.R (hunk 7 above)
Comment (BFalquet) [reply R/rules.R:178]: done.

File: R/rules.R [hunk 8]
```diff
@@ -125,3 +125,101 @@ as.list.rule <- function(x, ...) {
...
+#' )
+#'
+#' combine_list_rules(l1, l2)
+combine_list_rules <- function(x, val, ...) {
+  checkmate::assert_list(x, types = "rule", null.ok = FALSE, names = "unique")
+  checkmate::assert_list(val, types = "rule", null.ok = FALSE, names = "unique")
+
+  xnames <- names(x)
+  vnames <- names(val)
+  vnames <- vnames[nzchar(vnames)]
```
Comment (clarkliming) [inline R/rules.R:219]: names = "unique" should already stop on the non-zero character name

File: R/rules.R [hunk 9]
```diff
@@ -125,3 +125,101 @@ as.list.rule <- function(x, ...) {
...
+  names_x <- names(x)
+  names_x_diff <- setdiff(names_x, names_y)
+
+  x <- x[names_x_diff]
+  r <- c(y, x)
+
+  # Reorder to follow original order
+  names_y_diff <- setdiff(names_y, names_x)
+  names_order <- c(names_x, names_y_diff)
+  r <- r[names_order]
```
Comment (clarkliming) [inline R/rules.R:175]: current structure seems to be same as `modifyList(x, y)` or `x[names(y)] <- y` (which keeps the order of x)

**Review by clarkliming — APPROVED** [review]
looks good to me!

**Pull request merged**